File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-no-falsy-and-render.ts
import { defineRule } from "../../utils/define-rule.js";
import { hasDirective } from "../../utils/has-directive.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

const NUMERIC_NAME_HINTS = [
  "count",
  "length",
  "total",
  "size",
  "num",
  "index",
  "amount",
  "quantity",
  "offset",
  "width",
  "height",
  "duration",
  "progress",
  "score",
  "rank",
  "level",
  "step",
  "max",
  "min",
  "sum",
  "avg",
  "depth",
  "balance",
  "age",
  "weight",
  "volume",
  "distance",
  "speed",
  "rate",
  "ratio",
  "percent",
  "percentage",
];

const BOOLEAN_PREFIXES = [
  "is",
  "has",
  "can",
  "should",
  "did",
  "will",
  "show",
  "hide",
  "enable",
  "disable",
];

// HACK: word-boundary aware to avoid false positives like `discount`
// matching "count" or `isPage` matching "page".
const isNumericName = (name: string): boolean => {
  const lower = name.toLowerCase();
  for (const prefix of BOOLEAN_PREFIXES) {
    if (
      lower.startsWith(prefix) &&
      name.length > prefix.length &&
      name[prefix.length] === name[prefix.length].toUpperCase()
    ) {
      return false;
    }
  }

  for (const hint of NUMERIC_NAME_HINTS) {
    if (lower === hint) return true;
    const camelSuffix = hint.charAt(0).toUpperCase() + hint.slice(1);
    if (name.endsWith(camelSuffix)) return true;
    if (lower.endsWith(`_${hint}`)) return true;
  }
  return false;
};

const isLikelyNumericExpression = (node: EsTreeNode): boolean => {
  if (
    isNodeOfType(node, "MemberExpression") &&
    isNodeOfType(node.property, "Identifier") &&
    node.property.name === "length"
  )
    return true;

  if (isNodeOfType(node, "Identifier") && isNumericName(node.name)) return true;

  if (
    isNodeOfType(node, "MemberExpression") &&
    isNodeOfType(node.property, "Identifier") &&
    isNumericName(node.property.name)
  )
    return true;

  return false;
};

// HACK: `{count && <Component />}` renders the raw number `0` when
// `count` is 0. On React Native, rendering a bare number outside
// `<Text>` causes a hard production crash. This rule flags `&&`
// conditions that look like they could produce a numeric falsy value.
//
// We intentionally do NOT flag every `{x && <Y />}` — most are
// boolean state/props/constants that never produce `0`. We only
// flag identifiers/expressions with numeric-sounding names or
// `.length` access.
export const rnNoFalsyAndRender = defineRule<Rule>({
  id: "rn-no-falsy-and-render",
  title: "Numeric && renders bare zero",
  requires: ["react-native"],
  severity: "error",
  recommendation:
    "When the number is 0, this shows a bare `0` as text, which crashes on RN. Use `{value > 0 && <X />}`, `{Boolean(value) && <X />}`, or `{value ? <X /> : null}`.",
  create: (context: RuleContext) => {
    let isDomComponentFile = false;

    return {
      Program(programNode: EsTreeNodeOfType<"Program">) {
        isDomComponentFile = hasDirective(programNode, "use dom");
      },
      LogicalExpression(node: EsTreeNodeOfType<"LogicalExpression">) {
        if (isDomComponentFile) return;
        if (node.operator !== "&&") return;

        const isRightJsx =
          isNodeOfType(node.right, "JSXElement") || isNodeOfType(node.right, "JSXFragment");
        if (!isRightJsx) return;

        const parent = node.parent;
        const isInsideJsx =
          isNodeOfType(parent, "JSXExpressionContainer") ||
          (isNodeOfType(parent, "LogicalExpression") && parent.operator === "&&");
        if (!isInsideJsx) return;

        const left = node.left;
        if (!left) return;

        if (!isLikelyNumericExpression(left)) return;

        context.report({
          node: left,
          message: "Your users hit a crash when this value is 0 & renders a bare `0` as text.",
        });
      },
    };
  },
});

File: native/oxlint/rules/rn-no-falsy-and-render.rs
use oxc_ast::{
    AstKind,
    ast::{BindingPattern, Expression, JSXAttributeName, JSXElementName},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const NUMERIC_NAME_HINTS: [&str; 32] = [
    "count",
    "length",
    "total",
    "size",
    "num",
    "index",
    "amount",
    "quantity",
    "offset",
    "width",
    "height",
    "duration",
    "progress",
    "score",
    "rank",
    "level",
    "step",
    "max",
    "min",
    "sum",
    "avg",
    "depth",
    "balance",
    "age",
    "weight",
    "volume",
    "distance",
    "speed",
    "rate",
    "ratio",
    "percent",
    "percentage",
];
const BOOLEAN_PREFIXES: [&str; 10] = [
    "is", "has", "can", "should", "did", "will", "show", "hide", "enable", "disable",
];
const MESSAGE: &str = "Your users hit a crash when this value is 0 & renders a bare `0` as text.";

#[derive(Debug, Default, Clone)]
pub struct RnNoFalsyAndRender;

declare_oxc_lint!(
    /// Disallow numeric logical-and gates that render a bare zero on React Native.
    RnNoFalsyAndRender,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Numeric && renders bare zero.",
);

impl Rule for RnNoFalsyAndRender {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        is_react_native_file_active(ctx)
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if ctx.nodes().iter().any(|node| {
            matches!(node.kind(), AstKind::Program(program)
                if program.directives.iter().any(|directive| directive.expression.value == "use dom"))
        }) {
            return;
        }
        for node in ctx.nodes().iter() {
            let AstKind::LogicalExpression(logical) = node.kind() else {
                continue;
            };
            if logical.operator != oxc_syntax::operator::LogicalOperator::And
                || !matches!(
                    &logical.right,
                    Expression::JSXElement(_) | Expression::JSXFragment(_)
                )
            {
                continue;
            }
            let parent = ctx.nodes().parent_node(node.id());
            if !matches!(parent.kind(), AstKind::JSXExpressionContainer(_))
                && !matches!(parent.kind(), AstKind::LogicalExpression(parent_logical)
                    if parent_logical.operator == oxc_syntax::operator::LogicalOperator::And)
            {
                continue;
            }
            if rn_falsy_is_rendered_inside_dom_host(node, ctx)
                || !rn_falsy_is_likely_numeric(&logical.left)
            {
                continue;
            }
            if let Expression::Identifier(identifier) = &logical.left
                && rn_falsy_identifier_is_provably_safe(identifier, ctx)
            {
                continue;
            }
            ctx.diagnostic(OxcDiagnostic::error(MESSAGE).with_label(logical.left.span()));
        }
    }
}

fn rn_falsy_is_likely_numeric(expression: &Expression<'_>) -> bool {
    if let Expression::Identifier(identifier) = expression {
        return rn_falsy_is_numeric_name(identifier.name.as_str());
    }
    expression.as_member_expression().is_some_and(|member| {
        member_expression_identifier_property_name(member)
            .is_some_and(|name| name == "length" || rn_falsy_is_numeric_name(name))
    })
}

fn rn_falsy_is_numeric_name(name: &str) -> bool {
    let lowercase = name.to_ascii_lowercase();
    if BOOLEAN_PREFIXES.iter().any(|prefix| {
        lowercase.starts_with(prefix)
            && name
                .get(prefix.len()..)
                .and_then(|name| name.chars().next())
                .is_some_and(|character| character.to_uppercase().eq(std::iter::once(character)))
    }) {
        return false;
    }
    NUMERIC_NAME_HINTS.iter().any(|hint| {
        lowercase == *hint
            || lowercase.ends_with(&format!("_{hint}"))
            || name.ends_with(&format!("{}{}", hint[..1].to_ascii_uppercase(), &hint[1..]))
    })
}

fn rn_falsy_identifier_is_provably_safe<'a>(
    identifier: &oxc_ast::ast::IdentifierReference<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let Some(symbol_id) = ctx
        .scoping()
        .get_reference(identifier.reference_id())
        .symbol_id()
    else {
        return false;
    };
    let declaration = ctx.symbol_declaration(symbol_id);
    let AstKind::VariableDeclarator(declarator) = declaration.kind() else {
        return false;
    };
    let parent = ctx.nodes().parent_node(declaration.id());
    if matches!(parent.kind(), AstKind::VariableDeclaration(variable) if variable.kind.is_const())
        && declarator
            .init
            .as_ref()
            .is_some_and(rn_falsy_is_safe_initializer)
    {
        return true;
    }
    let BindingPattern::ArrayPattern(pattern) = &declarator.id else {
        return false;
    };
    if pattern
        .elements
        .first()
        .and_then(Option::as_ref)
        .and_then(BindingPattern::get_binding_identifier)
        .is_none_or(|binding| binding.symbol_id() != symbol_id)
    {
        return false;
    }
    let Some(Expression::CallExpression(call)) = &declarator.init else {
        return false;
    };
    call.callee_name() == Some("useState")
        && call
            .arguments
            .first()
            .and_then(oxc_ast::ast::Argument::as_expression)
            .is_some_and(rn_falsy_is_boolean_expression)
}

fn rn_falsy_is_safe_initializer(expression: &Expression<'_>) -> bool {
    if rn_falsy_is_boolean_expression(expression) {
        return true;
    }
    match expression {
        Expression::NumericLiteral(literal) => literal.value != 0.0,
        Expression::StringLiteral(literal) => !literal.value.is_empty(),
        _ => false,
    }
}

fn rn_falsy_is_boolean_expression(expression: &Expression<'_>) -> bool {
    match expression {
        Expression::BooleanLiteral(_) => true,
        Expression::UnaryExpression(unary) => {
            unary.operator == oxc_syntax::operator::UnaryOperator::LogicalNot
        }
        Expression::BinaryExpression(binary) => matches!(
            binary.operator,
            oxc_syntax::operator::BinaryOperator::Equality
                | oxc_syntax::operator::BinaryOperator::Inequality
                | oxc_syntax::operator::BinaryOperator::StrictEquality
                | oxc_syntax::operator::BinaryOperator::StrictInequality
                | oxc_syntax::operator::BinaryOperator::LessThan
                | oxc_syntax::operator::BinaryOperator::LessEqualThan
                | oxc_syntax::operator::BinaryOperator::GreaterThan
                | oxc_syntax::operator::BinaryOperator::GreaterEqualThan
        ),
        Expression::CallExpression(call) => call.callee_name() == Some("Boolean"),
        _ => false,
    }
}

fn rn_falsy_is_rendered_inside_dom_host<'a>(node: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
    for ancestor in ctx.nodes().ancestors(node.id()) {
        match ancestor.kind() {
            AstKind::JSXAttribute(attribute) => {
                if !matches!(&attribute.name, JSXAttributeName::Identifier(identifier)
                    if identifier.name == "children")
                {
                    return false;
                }
            }
            AstKind::JSXElement(element) => {
                if is_scoped_react_fragment_element(&element.opening_element.name, ctx) {
                    continue;
                }
                let JSXElementName::Identifier(identifier) = &element.opening_element.name else {
                    return false;
                };
                return crate::globals::HTML_TAG.contains(identifier.name.as_str())
                    || is_svg_tag_name(identifier.name.as_str());
            }
            AstKind::CallExpression(_)
            | AstKind::NewExpression(_)
            | AstKind::VariableDeclarator(_)
            | AstKind::AssignmentExpression(_)
            | AstKind::ObjectProperty(_)
            | AstKind::Function(_)
            | AstKind::ArrowFunctionExpression(_)
            | AstKind::Program(_) => return false,
            _ => {}
        }
    }
    false
}
